cmd/server/pactasrv/populate.go

```diff
@@ -0,0 +1,104 @@
+package pactasrv
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/RMI/pacta/oapierr"
+	"github.com/RMI/pacta/pacta"
+	"go.uber.org/zap"
+)
+
+func populateAll[Source any, TargetID ~string, Target any](
+	sources []*Source,
+	getTargetsFn func(*Source) ([]*Target, error),
+	getTargetIDFn func(*Target) TargetID,
+	lookupTargetsFn func([]TargetID) (map[TargetID]*Target, error),
+) error {
+	allTargets := []*Target{}
+	for i, source := range sources {
+		targets, err := getTargetsFn(source)
+		if err != nil {
+			return fmt.Errorf("getting %d-th targets: %w", i, err)
+		}
+		allTargets = append(allTargets, targets...)
+	}
+
+	seen := map[TargetID]bool{}
+	uniqueIds := []TargetID{}
+	for _, target := range allTargets {
+		id := getTargetIDFn(target)
+		if _, ok := seen[id]; !ok {
+			uniqueIds = append(uniqueIds, id)
+			seen[id] = true
+		}
+	}
+
+	populatedTargets, err := lookupTargetsFn(uniqueIds)
+	if err != nil {
+		return fmt.Errorf("looking up populated: %w", err)
+	}
+	for i, source := range sources {
+		targets, err := getTargetsFn(source)
+		if err != nil {
+			return fmt.Errorf("re-getting %d-th targets: %w", i, err)
+		}
+		for _, target := range targets {
+			id := getTargetIDFn(target)
+			if populated, ok := populatedTargets[id]; ok {
+				*target = *populated
+			} else {
+				return fmt.Errorf("can't find populated target %s", id)
+			}
+		}
+	}
+	return nil
+}
+
+func populatePortfoliosInPortfolioGroups(
+	s *Server,
```

Comment:
Since this function isn't generic, you should be able to define it on the server (i.e. `func (s *Server) populatePortfoliosInPortfolioGroups(...)`) instead of having to pass it in. I'm guessing this is the remnant of some refactoring.

And same elsewhere.

Reply:
Good catch!